Reorganization of Geometries

z3dev opened the pull request:
**WORK IN PROGRESS DO NOT MERGE**

The goal of this pull request is to complete the basic geometries. Each should be complete, 100% tested, reviewed, and ready for use in other functionality.

Geometries:
- geom2
  - [x] code complete
  - [x] tests complete, 100% coverage
  - [x] reviewed
- geom3
  - [x] code complete
  - [x] tests complete, 100% coverage
  - [ ] reveiwed
- path2
  - [x] code complete
  - [x] tests complete, 100% coverage
  - [ ] reveiwed
- poly3
  - [x] code complete
  - [ ] tests complete
  - [ ] reviewed

z3dev commented:
@kaosat-dev please review the code for geom2, geom3, path2.

i need to add some tests to obtain 100% coverage.

are you going to make the changes to poly3 on this branch?

z3dev commented:
> i need to add some tests to obtain 100% coverage.

The test suites for geom2, geom3, and path2 are complete. All have 100% coverage now. :)

kaosat-dev commented:
I will check things out tomorrow, thanks @z3dev !
No, any changes I will make will be after this, in another branch :) (keeping things simple & manageable)

z3dev commented:
Some things to think about...
- Consistency... The geometries should be small and concise. Anything else, like geom2.toOutlinePaths, should go into the operations. There's some incosistency with naming as well, i.e. reverse, flip, etc.
- Parameter checking... how much should be performed? see geom2.fromPoints()
- Canonicalize... I think we should drop the canonicalize of vec2, and use full float values. It's my belief that the precision issues are due to the reuse of planes, not the point values. Planes should never be reused.

kaosat-dev commented on src/geometry/geom2/canonicalize.js:
```diff
@@ -0,0 +1,24 @@
+ * @example
+ * let newgeometry = canonicalize(geometry)
+ */
+const canonicalize = (geometry) => {
```
nice and clean !

kaosat-dev commented on src/geometry/geom2/canonicalize.js:
```diff
@@ -0,0 +1,24 @@
+
+/**
+ * Produces a canonicalized geometry by canonicalizing the sides.
+ * Must be called before exposing any side data.
```
I think we should be more explicit about what canonicalizing means in the first place

z3dev replied:
I'm not a fan of the term. As of NOW, the logic just 'applies' the tranforms, so I suggest we change to...

- canonicalize() ==> applyTransforms()
- isCanonicalized ==> isTransformed

kaosat-dev replied:
I agree that the naming is better ! 
so we are clear on things
- raw geometry
- transform 
- applyTransform => sets the actual transforms, sets the 'isTransformed' flag

What happens if you apply multiple transforms ?

kaosat-dev replied:
@z3dev did you see my question above ? :)

z3dev replied:
Sorry.

See transforms(), as that function resets the isTransformed = false, and does the matrix multiplication.

In addition, it’s really important that the contents are accessed via toPolnts, toSide, or toPolygons. Those functions will apply the transforms and returns the contents.

z3dev replied:
@kaosat-dev Nice catch. There were some inconsistencies in the way transforms were being handled, as well as the isCanonicalized state. These were fixed. However, I would like to add test cases to make sure that the 'lazy' application REALLY works.

kaosat-dev replied:
ah sorry, I had missed that !
but that makes sense, also good idea on tests

kaosat-dev replied:
about numeric transformations and precision issues: we still need to check what happens inside of booleans and how 'very close' points are merged into a single one, but perhaps that should be another PR as it is a booleans issue

z3dev replied:
@kaosat-dev please review geom2 only.

i added a full scenario to test lazy transforms.

also, I found something very interesting. basically, the isCanonicalized boolean is unnecessary because the transforms are always reset to the indentity matrix. the transform() function only changes the transforms matrix, while the canonicalize() function applies the transforms then resets the transforms matrix.

z3dev replied:
@kaosat-dev please review geom2 again. how do those changes look to transform() and canonicalize() ?

kaosat-dev commented on src/geometry/geom2/canonicalize.test.js:
```diff
@@ -0,0 +1,31 @@
+test('canonicalize: Updates a populated geom2 with canonalized sides', (t) => {
+  const points = [[0, 0], [1, 0], [0, 1]]
+  const expected = {
+    baseSides: [
```
what are basesides ?

z3dev replied:
basePoints, baseSides, basePolygons have been removed.
i tested with the new primitives as well. :)

kaosat-dev commented on src/geometry/geom2/clone.js:
```diff
@@ -0,0 +1,8 @@
+ * @params {geom2} geometry - the geometry to make a shallow copy of
+ * @returns {geom2} new geometry
+ */
+const clone = (geometry) => Object.assign({}, geometry)
```
can you explain why clone is not explicit anymore ? Object.assign does not handle all copying as one would expect, so it needs to be used very carefully!

z3dev replied:
I depends on what we expect. Do we expect 'shallow' or 'deep' copies of the object?

math objects perform 'deep' copies.

kaosat-dev replied:
but should a clone not be deep by definition ? 
we never mutate things so I am not sure it matters in the end, but my point about the issues with Object.assign still stand sadly

z3dev replied:
Good. We just need to come to the same understanding.

I will change to be ‘deep’ clones.

z3dev replied:
Done. Clone performs DEEP copies of the contents. Please review.

kaosat-dev replied:
awesome ! I'll check the updated code

kaosat-dev replied:
ok seems good !

kaosat-dev commented on src/geometry/geom2/canonicalize.js:
```diff
@@ -0,0 +1,24 @@
+  if (geometry.isCanonicalized) return geometry
+
+  // apply transforms to each side, and canonicalize
+  geometry.sides = geometry.baseSides.map((side) => {
```
why basesides ?
storing the same (or almost same) data twice in immutable data structures does not make much sense
do not forget that we have in jscad :
- obj2 = operation(obj1)
- so we do not need to keep seperate structures (very costly copies) for sides that have been transformed (transformation here being canonicalized)

z3dev replied:
gone. please review again.

kaosat-dev commented on src/geometry/geom2/create.js:
```diff
@@ -0,0 +1,20 @@
+    sides = [] // empty contents
+  }
+  return {
+    baseSides : sides,
```
same point as above

kaosat-dev commented on src/geometry/geom2/create.js:
```diff
@@ -0,0 +1,20 @@
+    baseSides : sides,
+    sides : [],
+    isCanonicalized : false,
+    transforms : mat4.create()
```
so transforms are combined again with geom?

kaosat-dev commented on src/geometry/geom2/fromPoints.js:
```diff
@@ -0,0 +1,30 @@
+ * @returns {geom2} a new geometry
+ */
+const fromPoints = function (points) {
+  if (!Array.isArray(points)) {
```
nice

kaosat-dev commented on src/geometry/geom2/create.test.js:
```diff
@@ -0,0 +1,16 @@
+
+test('create: Creates an empty, uncanonicalized geom2', t => {
+  const expected = {baseSides: [], sides: [],
+                    isCanonicalized: false, transforms: new Float32Array([1,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1]) }
```
you can use a mat4.create instead of the manual floatArray right ? or did you want it explicit ?

kaosat-dev commented on src/geometry/geom2/isA.js:
```diff
@@ -0,0 +1,17 @@
+/**
+ * Determin if the given object is a 2D geometry.
```
Determine :)

kaosat-dev commented on src/geometry/geom2/isA.js:
```diff
@@ -0,0 +1,17 @@
+ * @params {geom2} object - the object to interogate
+ * @returns {true} if the object matches a geom2 based object
+ */
+const isA = (object) => {
```
we need something more clear I think; like  'isGeom2'

kaosat-dev commented on src/geometry/geom2/reverse.js:
```diff
@@ -0,0 +1,24 @@
+
+/**
+ * Reverses the given geometry so that the sides are flipped and in the opposite order.
+ * This swaps the left (interior) and right (exterior) edges.
```
I am not 100% sure what this means , does this change the winding order ?

kaosat-dev commented on src/geometry/geom2/toPoints.js:
```diff
@@ -0,0 +1,24 @@
+    return side[0]
+  })
+  // due to the logic of fromPoints()
+  // move the first point to the last
```
is this still needed in V2 ?

kaosat-dev commented on src/geometry/geom2/toPoints.test.js:
```diff
@@ -0,0 +1,26 @@
+  const points = [[0, 0], [1, 0], [0, 1]]
+  const geometry = fromPoints(points)
+
+  toString(geometry)
```
is this debug ?

kaosat-dev commented on src/geometry/geom2/toString.js:
```diff
@@ -0,0 +1,27 @@
+ */
+const toString = function (geometry) {
+  if (geometry.isCanonicalized) {
+    let result = 'geom2 (' + geometry.sides.length + ' sides):\n[\n'
```
please use es6 strings instead :)

kaosat-dev commented on src/geometry/geom2/transform.js:
```diff
@@ -0,0 +1,32 @@
+/**
+ * Transform the given geometry using the given matrix.
+ * This is a lazy transform of the sides, as this function only adjusts the transforms.
+ * See canonicalize() for the actual application of the transfroms to the sides.
```
wow ok, so with regard to my points above about baseSides: so canonicalize is needed to apply transforms ? I get the intention but that is very confusing to be honest.

kaosat-dev replied:
canonicalize was merging 'close enough' points, I do not see how applying transforms is related ?

kaosat-dev commented on src/geometry/path2/concat.js:
```diff
@@ -0,0 +1,30 @@
+ * @example
+ * let newpath = concat(fromPoints({}, [[1, 2]]), fromPoints({}, [[3, 4]]))
+ */
+const concat = (...paths) => {
```
I am scratching my head about this one, I'll get back to you on that :D

kaosat-dev commented on src/geometry/path2/close.js:
```diff
@@ -0,0 +1,34 @@
+ * @params {geometry} the path to close
+ * @returns {path} the closed path
+ */
+const close = (geometry) => {
```
sorry If I am being dense here, but why do we need so much complexity ?
Can a closed path not be defined just as having the same start & end point ?
And would closing a path not be just appending the point if needed ?

kaosat-dev commented on src/geometry/path2/reverse.js:
```diff
@@ -0,0 +1,22 @@
+ * @example
+ * reverse(path)
+ */
+const reverse = (path) => {
```
nice and simple !

kaosat-dev commented on src/geometry/poly3/clone.js:
```diff
@@ -0,0 +1,33 @@
+ * @param {vec3} poly3 - polygon to clone
+ * @returns {vec3} clone of the polygon
+ */
+const clone = (...params) => {
```
ok

kaosat-dev commented on src/geometry/poly3/create.js:
```diff
@@ -0,0 +1,45 @@
+ *
+ * @constructor
+ * @param {vec3[]} vertices - list of vertices
+ * @param {shared} [shared=defaultShared] - shared property to apply
```
nice, but shared is gone right?

kaosat-dev replied:
hmm shared was a generic data for color and more, perhaps I was wrong and we should keep it ?

kaosat-dev commented on src/geometry/poly3/fromPoints.js:
```diff
@@ -0,0 +1,27 @@
+ * ]
+ * const polygon = fromPoints(points)
+ */
+const fromPoints = (points, planeof) => {
```
planeOf => plane

kaosat-dev commented on src/geometry/poly3/fromPoints.js:
```diff
@@ -0,0 +1,27 @@
+ * const polygon = fromPoints(points)
+ */
+const fromPoints = (points, planeof) => {
+  if (planeof) throw new Error('use fromPointAndPlane')
```
nice throwings !

z3dev replied:
This is for development... I want to catch all invalid useage of this function.

kaosat-dev commented on src/geometry/poly3/fromPointsAndPlane.js:
```diff
@@ -0,0 +1,9 @@
+ * @param {Array[]} vertices - list of vertices
+ * @param {plane} [plane] - plane of the polygon
+ */
+const fromPointsAndPlane = (vertices, plane) => {
```
nice and specific

kaosat-dev commented on src/geometry/poly3/isA.js:
```diff
@@ -0,0 +1,15 @@
+ * @params {poly3} object - the object to interogate
+ * @returns {true} if the object matches a poly3 based object
+ */
+const isA = (object) => {
```
same remark as for others, although having a semi 'generic' operand name like this can be nice, not sure

z3dev replied:
i like this idea better then having a utilitiy function that does a determination. the object itself knows best.

kaosat-dev replied:
not sure I understand ?

z3dev replied:
There were functions called isCSG and isCAG in the utilities. I think these functions should be built into the objects, i.e. isA()

kaosat-dev replied:
ah yes I see your point !
hmm so we would have geom2.isA(xxx) that feels strange right ?
geom2.isOfThisType ? (meh)
isGeom2() feels more natural but you loose the reusability
then again clarity is better ?
not sure what naming is best to be honest

z3dev replied:
I kind of comes down to this...

Do we want uniquely named utility functions? isGeom2, isPath2, etc

OR

Do we want to use the namespace to call isA()? geom2.isA(), path2.isA(), etc.

You might be happier with names like geom2.isCompatible() or geom2.isEquivalentType() or geom2.isUsableAs()

kaosat-dev commented on src/geometry/poly3/measureBoundingSphere.js:
```diff
@@ -0,0 +1,15 @@
+const vec3 = require('../../math/vec3')
```
I am starting to think that we could have some generic measurement functions ?
might be premature, but could we not just operate on the underlying points ?

z3dev replied:
nice idea. it would be possible now because all the geometries are point based. we will see.

kaosat-dev commented on src/geometry/geom2/applyTransforms.js:
```diff
@@ -2,14 +2,14 @@ const mat4 = require('../../math/mat4')
- * @param {geom2} geometry - the geometry to canonicalize
- * @returns {geom2} the given geometry with transformed points
+ * Apply the transforms of the given geometry.
+ * NOTE: Must be called before exposing any side data.
```
I would change the note to something more clear like 'call this first if you need the side's data with transforms baked in'

kaosat-dev replied:
also perhaps mention, to sides ?

kaosat-dev commented on src/geometry/geom2/toSides.js:
```diff
@@ -9,7 +9,7 @@ const canonicalize = require('./canonicalize')
  * let sharedsides = toSides(geometry)
  */
 const toSides = function (geometry) {
```
ahh smart !

kaosat-dev replied:
I really like this, makes sides data independant from data structures, well done

z3dev commented:
@kaosat-dev Please review geom3 and path2, as these have been updated.
- renamed canonicalize to applyTransforms
- removed isCanonicalized as the transforms can be used to determine state
- updated all logic and tests

z3dev commented:
> @kaosat-dev Please review geom3 and path2, as these have been updated.

@kaosat-dev Can you review and merge this pull request?

If any changes are required then let me know. Cheers

kaosat-dev commented:
this all looks amazing ! thanks @z3dev  going to merge now !
